go-controller/pkg/kube/annotator.go

```diff
@@ -14,6 +15,8 @@ type Annotator interface {
 	SetWithFailureHandler(key string, value interface{}, failFn FailureHandlerFn) error
 	Delete(key string)
 	Run() error
+	Lock()
```

Comment:
@trozet I don't think we need to export these?

Reply:
if we want type Annotator to always be thread safe, it needs to implement Lock and Unlock from Mutex to ensure someone doesnt create an unsafe type of Annotator

Reply:
Yeah, but we never want things outside the Annotator to use them. it's totally breaking encapsulation.

Reply:
So what you do is something like this:

```
diff --git a/go-controller/pkg/kube/annotator.go b/go-controller/pkg/kube/annotator.go
index ba2dd11454166..24d1bfd95c8b2 100644
--- a/go-controller/pkg/kube/annotator.go
+++ b/go-controller/pkg/kube/annotator.go
@@ -4,6 +4,7 @@ import (
 	"encoding/json"
 	"fmt"
 	"reflect"
+	"sync"
 
 	kapi "k8s.io/api/core/v1"
 )
@@ -15,6 +16,12 @@ type Annotator interface {
 	Run() error
 }
 
+type threadSafeAnnotator interface {
+	Annotator
+	Lock()
+	Unlock()
+}
+
 type action struct {
 	key     string
 	val     string
@@ -22,14 +29,20 @@ type action struct {
 }
 
 type nodeAnnotator struct {
+	sync.Mutex
 	kube Interface
 	node *kapi.Node
-
 	changes map[string]*action
 }
 
+var _ threadSafeAnnotator = &nodeAnnotator{}
+
 // NewNodeAnnotator returns a new annotator for Node objects
 func NewNodeAnnotator(kube Interface, node *kapi.Node) Annotator {
+	return newNodeAnnotator(kube, node)
+}
+
+func newNodeAnnotator(kube Interface, node *kapi.Node) threadSafeAnnotator {
 	return &nodeAnnotator{
 		kube:    kube,
 		node:    node,
@@ -86,6 +99,10 @@ func (na *nodeAnnotator) Run() error {
 
 // NewPodAnnotator returns a new annotator for Pod objects
 func NewPodAnnotator(kube Interface, pod *kapi.Pod) Annotator {
+	return newPodAnnotator(kube, pod)
+}
+
+func newPodAnnotator(kube Interface, pod *kapi.Pod) threadSafeAnnotator {
 	return &podAnnotator{
 		kube:    kube,
 		pod:     pod,
@@ -94,12 +111,14 @@ func NewPodAnnotator(kube Interface, pod *kapi.Pod) Annotator {
 }
 
 type podAnnotator struct {
+	sync.Mutex
 	kube Interface
 	pod  *kapi.Pod
-
 	changes map[string]*action
 }
 
+var _ threadSafeAnnotator = &podAnnotator{}
+
 func (pa *podAnnotator) Set(key string, val interface{}) error {
 	act := &action{
 		key:     key,
@@ -149,6 +168,10 @@ func (pa *podAnnotator) Run() error {
 
 // NewNamespaceAnnotator returns a new annotator for Namespace objects
 func NewNamespaceAnnotator(kube Interface, namespace *kapi.Namespace) Annotator {
+	return newNamespaceAnnotator(kube, namespace)
+}
+
+func newNamespaceAnnotator(kube Interface, namespace *kapi.Namespace) threadSafeAnnotator {
 	return &namespaceAnnotator{
 		kube:      kube,
 		namespace: namespace,
@@ -157,12 +180,14 @@ func NewNamespaceAnnotator(kube Interface, namespace *kapi.Namespace) Annotator
 }
 
 type namespaceAnnotator struct {
+	sync.Mutex
 	kube      Interface
 	namespace *kapi.Namespace
-
 	changes map[string]*action
 }
 
+var _ threadSafeAnnotator = &namespaceAnnotator{}
+
 func (na *namespaceAnnotator) Set(key string, val interface{}) error {
 	act := &action{
 		key:     key,
```

Reply:
yeah, the interface is an abstraction, the locks are part of the implementaton

Reply:
We came up with a better implementation that (a) enforces thread-safety via the interface and (b) doesn't export the lock/unlock functions.

Reply:
unfortunately it didnt work due to export problems:
```
Cannot use 'fakeAnnotator' (type *mocks.Annotator) as type kube.Annotator Type cannot implement 'kube.Annotator' as it has a non-exported method and is defined in a different package
```

going to just put a comment on the interface to say you should implement with thread safety